### ISSUE TITLE
Sinon breaks on import

### ISSUE THREAD

**Issue by domderen** [sinonjs/sinon#758]
Hi,

I have a project that uses sinon in unit tests. The project worked just fine could of days ago, and during that time I didn't change anything in it. Today I cloned it on a different machine, and tried running the tests again, but they now break with stack trace:

```
/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_xdomain_request.js:13
    undefined.sinon = {};
                    ^
TypeError: Cannot set property 'sinon' of undefined
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_xdomain_request.js:13:15)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at loadDependencies (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server.js:228:9)
    at /Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server.js:238:9
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server.js:242:2)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at loadDependencies (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server_with_clock.js:89:9)
    at /Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server_with_clock.js:97:9
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/util/fake_server_with_clock.js:101:2)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at loadDependencies (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/sandbox.js:147:9)
    at /Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/sandbox.js:156:9
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon/sandbox.js:162:2)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at loadDependencies (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon.js:28:9)
    at /Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon.js:39:9
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/sinon/lib/sinon.js:46:2)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/test/unit/ValidationUtil-test.js:2:46)
    at Module._compile (module.js:431:26)
    at normalLoader (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:160:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/dominik/dev/testing-presentation/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:173:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at /Users/dominik/dev/testing-presentation/node_modules/mocha/lib/mocha.js:192:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/dominik/dev/testing-presentation/node_modules/mocha/lib/mocha.js:189:14)
    at Mocha.run (/Users/dominik/dev/testing-presentation/node_modules/mocha/lib/mocha.js:422:31)
    at Object.<anonymous> (/Users/dominik/dev/testing-presentation/node_modules/mocha/bin/_mocha:398:16)
    at Module._compile (module.js:431:26)
    at Object.Module._extensions..js (module.js:449:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:472:10)
    at startup (node.js:124:18)
    at node.js:959:3
```

The project I speak about is available at: https://github.com/domderen/testing-presentation

It looks to me like some sinon dependency got updated and it now breaks sinon during requireing. I'd appreciate any help on this one.

**Comment by joakimbeng** [comment]
I'm getting the same issue (on node@0.12).
I think it's the combination of `"use strict"` and the usage of `this` that's causing the problem.
If I change line 13 and 223 to use `global` instead of `this` it's working, but then it won't work in a browser.

To be on the safe side maybe a function like the following should be put in the top of the file:

```javascript
function getGlobal() {
    return typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : this;
}
```

And then use that on line 13 and 223 like so:

```javascript
if (typeof sinon == "undefined") {
    getGlobal().sinon = {}; // L13
}
// ...
})(getGlobal()); // L223
```

What do you think?

**Comment by domderen** [comment]
The issue seems to be fixed now. It magically started to work again.

**Comment by rbirkby** [comment]
This happened for me immediately after I began using gulp with built-in babel support. As soon as the babel/register module was loaded, my tests fail with the same stacktrace as above.

**Comment by djMax** [comment]
This is happening to me now... deleting and reinstalling doesn't help it. Works fine on my Mac, but not on our Ubuntu CI server

**Comment by djMax** [comment]
I don't understand all the details yet - but I'm pretty sure this happens when babel decides to babelify sinon. So with appropriate babel register configuration, I got it to go away.

**Comment by ThomasConner** [comment]
@djMax I am running into the same issue. What babel register configuration did you use to make this go away? I am invoking my tests running the command `mocha --compilers js:babel/register --reporter spec --slow 500 --timeout 2000 test/specs/user.spec.js`

**Comment by eguneys** [comment]
https://github.com/badsyntax/react-seed/issues/21

**Comment by yanivtal** [comment]
Is anyone looking at this? Sinon doesn't work with ES6. You're gonna miss the next crop of developers!

**Comment by mroderick** [comment]
@yanivtal It's open source, why don't you help out if you're so concerned about the next crop of developers?

**Comment by Serfenia** [comment]
I currently have this issue when using native ES6 classes together with 'use strict'.  Changing the code on line 13 and 223 as @joakimbeng said does the trick for me. Any chance this can be fixed soon?

**Comment by fatso83** [comment]
@Serfenia : you are welcome to fix it.

The problem is that I have not seen a consistent way of reproducing this from a clean slate. If anyone can submit a gist that when executing will display this problem in the latest version of Sinon, this has a much higher chance of being fixed. If it is indeed still an issue, that is. No way of telling without a test case.

**Comment by Serfenia** [comment]
I made a small project that reproduces the issue. Check it out: [sinon-use-strict-native](https://github.com/Serfenia/sinon-use-strict-native)

**Comment by fatso83** [comment]
That's great, @Serfenia! Thanks.

**Comment by Serfenia** [comment]
@fatso83 I just pulled master to check if my fix does not break any tests within the code, but the structure of the master branch is different from the one on npm, is that correct? Which branch equals the npm package in it's entirety structure and file wise?

**Comment by fatso83** [comment]
That's great news. You are correct about the branches: master is now in preparation for the next major release (version 2), and so we are restructuring it to be able to use Browserify. The v1.17 branch is where we fix things like this that applies to the current version in NPM.

**Comment by fatso83** [comment]
@domderen This should be fixed in the v1.17 branch with the fixes from @Serfenia . Care to test it?

**Comment by domderen** [comment]
@fatso83 As I wrote on June 5, I am no longer able to reproduce the issue. Maybe someone who reproduces it?

**Comment by Serfenia** [comment]
@domderen check out my [sinon-use-strict-native](https://github.com/Serfenia/sinon-use-strict-native) branch. It reproduces the issue.

**Comment by fatso83** [comment]
I used the sinon-use-strict-native project of @Serfenia to verify that this indeed has been fixed in both `master` and `v1.17` (although I had to uncomment the `ignore` statement in `babelhook.js` to make it run). Closing as fixed
